src/index.cpp

```diff
@@ -1509,17 +1841,93 @@ namespace diskann {
                     << scratch->get_L() << std::endl;
     }
 
+    std::vector<LabelT> dummy;
     std::vector<unsigned> init_ids;
     init_ids.push_back(_start);
     std::shared_lock<std::shared_timed_mutex> lock(_update_lock);
-    auto retval =
-        iterate_to_fixed_point(query, L, init_ids, scratch, true, true);
+
+    auto retval = iterate_to_fixed_point(query, L, init_ids, scratch, false,
+                                         dummy, true, true);
+
     NeighborPriorityQueue &best_L_nodes = scratch->best_l_nodes();
 
     size_t pos = 0;
     for (int i = 0; i < best_L_nodes.size(); ++i) {
       if (best_L_nodes[i].id < _max_points) {
-        // safe because Index uses uint32_t ids internally 
+        // safe because Index uses uint32_t ids internally
+        // and IDType will be uint32_t or uint64_t
+        indices[pos] = (IdType) best_L_nodes[i].id;
+        if (distances != nullptr) {
+#ifdef EXEC_ENV_OLS
+          // DLVS expects negative distances
+          distances[pos] = best_L_nodes[i].distance;
+#else
+          distances[pos] = _dist_metric == diskann::Metric::INNER_PRODUCT
+                               ? -1 * best_L_nodes[i].distance
+                               : best_L_nodes[i].distance;
+#endif
+        }
+        pos++;
+      }
+      if (pos == K)
+        break;
+    }
+    if (pos < K) {
+      diskann::cerr << "Found fewer than K elements for query" << std::endl;
+    }
+
+    return retval;
+  }
+
+  template<typename T, typename TagT, typename LabelT>
+  template<typename IdType>
+  std::pair<uint32_t, uint32_t> Index<T, TagT, LabelT>::search_with_filters(
+      const T *query, const LabelT &filter_label, const size_t K,
+      const unsigned L, IdType *indices, float *distances) {
+    if (K > (uint64_t) L) {
+      throw ANNException("Set L to a value of at least K", -1, __FUNCSIG__,
+                         __FILE__, __LINE__);
+    }
+
+    ScratchStoreManager<InMemQueryScratch<T>> manager(_query_scratch);
+    auto                                      scratch = manager.scratch_space();
+
+    if (L > scratch->get_L()) {
+      diskann::cout << "Attempting to expand query scratch_space. Was created "
+                    << "with Lsize: " << scratch->get_L()
+                    << " but search L is: " << L << std::endl;
+      scratch->resize_for_new_L(L);
+      diskann::cout << "Resize completed. New scratch->L is "
+                    << scratch->get_L() << std::endl;
+    }
+
+    std::vector<LabelT>    filter_vec;
```

Comment:
Pre-allocate these containers in thread specific scratch

Reply:
This requires templating InMemScratch (as it does not know of LabelT) seems to be difficult and the related changes do bleed outside the scope and into other parts of the project. I think a good time to resolve this will be when we make LabelT globally as uint32.